ppoprf/src/ppoprf.rs

```diff
@@ -125,53 +136,47 @@ where
 pub struct Server {
     oprf_key: Scalar,
     public_key: ServerPublicKey,
-    mds: Vec<Vec<u8>>,
+    mds: Vec<u8>,
     pprf: GGM,
 }
 impl Server {
-    pub fn new(mds: &[Vec<u8>]) -> Result<Self, PPRFError> {
+    pub fn new(mds: Vec<u8>) -> Result<Self, PPRFError> {
         let mut csprng = OsRng;
         let oprf_key = Scalar::random(&mut csprng);
-        let mut public_key = Vec::with_capacity(mds.len() + 1);
-        public_key.push(oprf_key * RISTRETTO_BASEPOINT_POINT);
+        let mut md_pks = HashMap::new();
         let pprf = GGM::setup();
-        for md in mds {
+        for &md in mds.iter() {
             let mut tag = [0u8; 32];
-            pprf.eval(md, &mut tag)?;
+            pprf.eval(&[md], &mut tag)?;
             let ts = Scalar::from_bytes_mod_order(tag);
-            public_key.push(ts * RISTRETTO_BASEPOINT_POINT);
+            md_pks.insert(md, ts * RISTRETTO_BASEPOINT_POINT);
         }
         Ok(Self {
             oprf_key,
-            public_key,
-            mds: mds.to_vec(),
+            public_key: ServerPublicKey {
+                base_pk: oprf_key * RISTRETTO_BASEPOINT_POINT,
+                md_pks,
+            },
+            mds,
             pprf,
         })
     }
 
-    pub fn eval(
-        &self,
-        p: &Point,
-        md_idx: usize,
-        verifiable: bool,
-    ) -> Result<Evaluation, PPRFError> {
+    pub fn eval(&self, p: &Point, md: u8, verifiable: bool) -> Result<Evaluation, PPRFError> {
         let p = p.0;
         let point = p.decompress().unwrap();
-        if md_idx >= self.mds.len() {
-            return Err(PPRFError::BadTagIndex {
-                index: md_idx,
-                tag_size: self.mds.len(),
-            });
+        if !self.mds.iter().any(|&md_stored| md_stored == md) {
```

Comment:
Would `self.mds.get(md).is_none()` work?

Reply:
yep, that would also work 👍